### ISSUE TITLE
Document with ID containing what looks like a hex escape cannot be read

### ISSUE THREAD

**Issue by Liversage** [Azure/azure-documentdb-dotnet#458]
If the ID of a document contains the percent character (`%`) it may sometimes be impossible to read that document.

Here is a code example that creates a document with Id `%d5` and then tries to read it:

    var databaseId = "Test";
    var documentCollectionId = "Test";
    var documentId = "%d5";

    var client = new DocumentClient(new Uri("https://localhost:8081/"), "C2y6yDjf5/R+ob0N8A7Cgv30VRDJIWEHLM+4QDU5DE2nQ9nDuVTqobD4b8mGGyPMbIZnqyMsEcaGQy67XIw/Jw==");
    var database = (await client.CreateDatabaseIfNotExistsAsync(new Database { Id = databaseId })).Resource;
    var documentCollection = (await client.CreateDocumentCollectionIfNotExistsAsync(database.SelfLink, new DocumentCollection { Id = documentCollectionId })).Resource;

    var document = new Document { Id = documentId };
    await client.CreateDocumentAsync(documentCollection.SelfLink, document);

    var documentUri = UriFactory.CreateDocumentUri(databaseId, documentCollectionId, documentId);
    var response = await client.ReadDocumentAsync(documentUri);

The document is created and can be found using the document explorer with the correct key. However, an exception is thrown in the last line of the code:

> The input authorization token can't serve the request. Please check that the expected payload is built as per the protocol, and check the key being used. Server used the following payload to sign: 'get
> docs
> dbs/Test/colls/Test/docs/Õ
> thu, 22 mar 2018 23:38:08 gmt
>
> ActivityId: 64c382c7-1ae0-48de-bddf-6178248c4cb6, Microsoft.Azure.Documents.Common/1.21.0.0, documentdb-dotnet-sdk/1.21.1 Host/32-bit MicrosoftWindowsNT/10.0.16299.0

If you look carefully you will see an odd capital O with tilde character (`Õ`) in the exception message. To include this character in a URI it will normally be escaped as `%d5` which explains the link between the document ID and this odd character. However, if you examine `documentUri` it is `dbs/Test/colls/Test/docs/%25d5` so it is correctly escaped and as far as I can see this must be a bug in the SDK.

I get this error using the simulator (as in the code above) and also using a real Cosmos DB instance. I am using NuGet package version 1.21.1.

It seems that this problem only occurs if the percent sign is followed by exactly two hex digits. If the document ID is `%00` you get another error message but I am pretty it is the same underlying problem.

**Comment by kirankumarkolli** [comment]
@Liversage thanks for reporting it. We will root-cause it and fix it. 

As a work-around you can use below ending which covers all scenarios 
Encode: Convert.ToBase64String(System.Text.Encoding.UTF8.GetBytes(id)).Replace("/", "-")
Decode: Encoding.UTF8.GetString(Convert.FromBase64String(id.Replace("-", "/"))) 

Note: Base64 has '/' and last Replace is to cover it with non-base64 valid character.

**Comment by FelschR** [comment]
I've just encountered this issue. Spent a while debugging everything, wondering why I get a Document not found result.

Either DocumentDB shouldn't accept document creation with IDs like this or it should support querying them without an issue.

**Comment by zbynek001** [comment]
Any news on this?
I've many documents having e.g. %252F inside Id, which cannot be accessed directly via documentUri, e.g. ReplaceDocumentAsync. Only workaround is to query them via "select * from c where c.id = 'XXX'" and the use selflink. And since we're not using any disallowed character inside Id, we expect this to work without issues.
All those documents can be accessed via data explorer

**Comment by jorgensigvardsson** [comment]
@zbynek001 Can you really access the documents in the document explorer? I had no such luck. The documents showed up in the document list, but errors occurred when clicking on the documents.

This is silly. I have strings naturally containing slashes, but I cannot use them in id's. Smells like abstraction leakage (REST compatbility anyone?). This means that I'll have to escape the strings from within the application. I did try to use auto generated IDs, put the "real ID" in a property of its own along with a unique key. Unfortunately, that unique key must be paired with the partition key, otherwise I can't get direct access to the documents. I really don't want to query my documents, as that consumes RUs like crazy.

Please work on removing this limitation from IDs!

**Comment by jorgensigvardsson** [comment]
By the way, this bug/limitation is in the v3 of the SDK as well.

**Comment by Liversage** [comment]
@zbynek001 If I understand your problem correctly you want to use slash (`/`) in your ID. However, that is not allowed. The [Resouce ID Property](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.resource.id?view=azure-dotnet
) has the following statement:

> The following characters are restricted and cannot be used in the Id property: '/', '\\', '?', '#'

This is a limitation of Cosmos DB and has nothing to do with the .NET SDK. I believe it is because URLs are used to address Cosmos DB resources.

However, the issue I have reported is a specific bug in the v2 .NET SDK and there is no problem accessing documents with the problematic ID using the document explorer. I have also tested that the v3 SDK does not have this bug.

Because certain characters are restricted (see above) and other characters are problematic/buggy (`%`) I encode all my Cosmos DB IDs to avoid any errors. For example the natural key `my/id` is converted to the Cosmos DB ID `my@2fid`.

**Comment by zbynek001** [comment]
Not exactly. I don't want to use "/" inside Id. Instead, I want to use e.g. "%2F" or "%26" or other url encoded characters inside Id. And as you can see, "%" is not in the list of restricted characters, but still doesn't work...

**Comment by kirankumarkolli** [comment]
@zbynek001 that's true. Its a bug on our side. One work-around is to do custom encoding (ex: varient of base64 excluding unsupported characters).

**Comment by codecadwallader** [comment]
This looks to still be an issue in v3 of the SDK as well.  Anyone know if it's being worked/addressed or if there is a workaround proposed?  I've got tens of millions of documents I can't delete now.. 😬

**Comment by vsadams** [comment]
I think the problem stems from the DocumentServiceRequest class inside of the Microsoft.Azure.Cosmos.Direct library. 

It looks like its using the PathsHelper.TryParsePathSegmentsWithDatabaseAndCollectionNames method to create the resource address based on the out argument resourceIdOrFullName.

The problem is that the helper method will unescape the data string. 
I think this can be fixed by modifying the following code block.

```
if (isNameBased)
{
    this.ResourceAddress = resourceIdOrFullName;
    this.DatabaseName = databaseName;
    this.CollectionName = collectionName;
}
```

to

``` 
if (isNameBased)
{
    this.ResourceAddress = Uri.EscapeUriString(resourceIdOrFullName);
    this.DatabaseName = databaseName;
    this.CollectionName = collectionName;
}
```

I cant submit a pr to that lib bc its not open source otherwise I would. I could be wrong but I believe that would fix it reading the code.

**Comment by vsadams** [comment]
Is there any update on this issue?

**Comment by codecadwallader** [comment]
Another general (if tedious) workaround is to create stored procedures.  Server side JavaScript bypasses the SDK bug.  If you already have existing documents and it's too late to apply a custom encoding scheme this may be helpful.

@kirankumarkolli Any update on this issue?
